### PR TITLE
Keep device language when it has been previosuly set, after a successful delegated authentication flow that clears localStorage

### PR DESCRIPTION
**Pull request by mgcm**
EW handles [language setting on init](https://github.com/vector-im/element-web/blob/e6f69f7dce051e9435b6bb81d8b8b74857d7d2cc/src/vector/init.tsx#L133) and saves it in localStorage, using React SDK's `SettingsStore`.

When in a delegated auth environment, the [success handler will clear the localStorage](https://github.com/matrix-org/matrix-react-sdk/blob/6b73700b051b607c950e9dfc88de08100222bdb9/src/Lifecycle.ts#L407), including the previously set local settings language value.

This results in not having that language value available in localStorage right after starting a new user session within such an environment.

This PR fixes that by checking if a language value is set in the `SettingsStore` and keeping it after clearing the localStorage.

## Checklist

-   [X] Tests written for new code (and old code if feasible)
-   [X] Linter and other CI checks pass
-   [X] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

Type: defect

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Keep device language when it has been previosuly set, after a successful delegated authentication flow that clears localStorage ([\#11902](https://github.com/matrix-org/matrix-react-sdk/pull/11902)). Contributed by @mgcm.

### CODE REVIEW

File: src/Lifecycle.ts [hunk 1]
```diff
@@ -1105,6 +1106,9 @@ export async function onLoggedOut(): Promise<void> {
  */
 async function clearStorage(opts?: { deleteEverything?: boolean }): Promise<void> {
     if (window.localStorage) {
+        // save the language code, if it has already been set
+        const language = SettingsStore.getValue("language");
```
Comment (t3chguy) [inline src/Lifecycle.ts:1110]: This comment doesn't align with what it does, as getValue will fall through to defaults unless `excludeDefaults` is specified

image

This is a fresh incognito window where I did not set language for example.

File: src/Lifecycle.ts [hunk 2]
```diff
@@ -1118,8 +1122,12 @@ async function clearStorage(opts?: { deleteEverything?: boolean }): Promise<void
             logger.error("idbDelete failed for account:mx_access_token", e);
         }
 
-        // now restore those invites and registration time
+        // now restore those invites, registration time and previously set language
         if (!opts?.deleteEverything) {
+            if (language) {
```
Comment (t3chguy) [inline src/Lifecycle.ts:1127]: this if check is redundant, language will always be set due to no excludeDefault

**Review by t3chguy — CHANGES_REQUESTED** [review]
The way this PR stands, it'll bake the default (config.json) language into localStorage at the time the code runs, this is not expected. The intent is to preserve the setting, but its also "preserving" the default into a Device-level setting.

**Comment by mgcm** [comment]
@t3chguy Would you agree to check directly within `mx_local_settings` for the existence of the `language` key / value instead ?

**Comment by t3chguy** [comment]
@mgcm no, take a look at https://github.com/vector-im/element-web/blob/0040181489feda0e5d70becdea8585ac5a344d99/src/vector/init.tsx#L122

**Comment by mgcm** [comment]
@t3chguy changed to get the value at the device level exclusively

**Comment by t3chguy** [comment]
@mgcm I'm not seeing any new commits from you, once you wish for a re-review please request one top right

**Review by t3chguy — COMMENTED** [review]
This now looks like it does the right thing, any chance of tests?

File: test/components/structures/MatrixChat-test.tsx [hunk 3]
```diff
@@ -225,6 +225,272 @@ describe("<MatrixChat />", () => {
...
+                );
+
+                // client successfully started
+                expect(defaultDispatcher.dispatch).toHaveBeenCalledWith({ action: "client_started" });
+
+                // check we get to logged in view
+                await waitForSyncAndLoad(loginClient, true);
+            });
+
+            it("should persist device language when available", async () => {
```
Comment (mgcm) [inline test/components/structures/MatrixChat-test.tsx:479]: this is the new test

**Comment by mgcm** [comment]
I struggled a bit to find the most appropriate place for adding the test and settled with the `MatrixChat-test` suite because it's where auth flows are being tested.

I also moved the OIDC tests up due to having the new test case failing when it runs after the "with an existing session / logout", which seem to mess up the test state (due to the actual dispatch logout action).

Would appreciate some feedback :)

File: test/components/structures/MatrixChat-test.tsx [hunk 4]
```diff
@@ -225,6 +225,272 @@ describe("<MatrixChat />", () => {
...
+                localStorage.setItem("mx_local_settings", localSettings);
+
+                jest.spyOn(Lifecycle, "attemptDelegatedAuthLogin");
+
+                getComponent({ realQueryParams });
+                await flushPromises();
+
+                expect(Lifecycle.attemptDelegatedAuthLogin).toHaveBeenCalled();
+                expect(localStorage.getItem("mx_local_settings")).toEqual(localSettings);
+            });
```
Comment (t3chguy) [inline test/components/structures/MatrixChat-test.tsx:489]: This tests the happy path - great - it'd also be good to test the path which the original solution did not handle - not setting the language if there wasn't one set to begin with, please.

Comment (mgcm) [reply test/components/structures/MatrixChat-test.tsx]: Added the unhappy path and slight refactor. Thanks!

**Review by t3chguy — APPROVED** [review]
Looks good to me, thanks.

The type fail is unrelated and has a PR pending review to fix

**Comment by t3chguy** [comment]
@mgcm if you could pull in latest develop that should fix CI

**Pull request merged**